### ISSUE TITLE
MinSR regularisation

### ISSUE THREAD

**Issue by attila-i-szabo** [netket/netket#1696]
I've just seen [this paper](https://arxiv.org/pdf/2401.10190.pdf). The algorithm they describe is interesting too, and we might want to consider supporting it, but there's a more immediate point. In Sec. 3.2, they explain that the MinSR matrix is guaranteed to have (1,1,...,1) as a zero vector, so regularising it (a lot) is a good idea to protect from numerical instability (in exact arithmetic, adding any multiple of the all-ones matrix makes no difference).

Does NetKet already do something similar?

**Comment by PhilipVinc** [comment]
I was discussing it this morning as well. I think their algorithm is resumed well in this figure

<img width="798" alt="Screenshot 2024-01-19 at 10 36 21" src="https://github.com/netket/netket/assets/2407108/bac12ee3-8763-4b94-a503-8ce438f4660c">


The major difference is line 8 of the algorithm, which requires replacing the local energies with something that depends on the previous gradient. 
This can easily be implemented in another driver by combining `local energies` and `jacobian` in [here](https://github.com/netket/netket/blob/bb86b0234d2941f9cb4e8cf6d6afc939c8d722f1/netket/experimental/driver/vmc_srt.py#L253).

Also, a little detail, line 9 they also add a full matrix of `1/Ns` for increased numerical stability. Unsure if this helps in general...
Then, they add momentum and gradient clipping which is easy to do with optax.

**Comment by PhilipVinc** [comment]
cc @riccardo-rende @llviteritti

**Comment by attila-i-szabo** [comment]
> Also, a little detail, line 9 they also add a full matrix of `1/Ns` for increased numerical stability. Unsure if this helps in general...

Yeah, I was mostly talking about this - it should improve MinSR too, and seems like a quick change there to someone who's worked on that code

**Comment by llviteritti** [comment]
Probably the easiest thing is to implement it in another driver starting from `VMC_SRt` and taking into account line 8 and then as @PhilipVinc said add moment and gradient clipping with optax. It would be interesting to see whether in a model on lattice it actually gives an improvement with respect to vanilla MinSR ..

**Comment by attila-i-szabo** [comment]
> Then, they add momentum and gradient clipping which is easy to do with optax.

Is this something (especially momentum) that we can encapsulate into a new driver? The updates would be wrong without adding the momentum, so it wouldn't make sense to leave it to the user to add it with an external momentum driver in optax (especially that conventions differ, so it would be an invitation for errors). Personally, I also don't see the point for using a separate library for this particular line of the algorithm, which is basically a single FMA.

Gradient clipping is a different story, I think it can be left up to users whether they like it and do it themselves in optax.

**Comment by hendrydouglas11** [comment]
The addition of the projector P = (1/N) one_vec * one_vec^T should be scaled since O*O^T -> O*O^T + P makes one_vec go from an eigen value 0 -> 1, but if the other eigenvalues of O*O^T are all much less than 1, then it becomes the dominate eigen value and can leave the matrix still ill conditioned (going from one extreme to the other).  I instead rescaled by tr( O*O^T) /N =avg(eig(O*O^T)) which ensures that it becomes neither the largest nor smallest eigen value.  This worked well for me, also the Kaczmarz/momentum parts did make a noticeably improvement over minSR for 1D Heisenberg with RBMs.

**Comment by chrisrothUT** [comment]
> In Sec. 3.2, they explain that the MinSR matrix is guaranteed to have (1,1,...,1) as a zero vector, so regularising it (a lot) is a good idea to protect from numerical instability (in exact arithmetic, adding any multiple of the all-ones matrix makes no difference).

@attila-i-szabo this is why it's better to regularize MinSR by ignoring the modes where the eigenvalues are below some threshold rather than adding a diag_shift. Just throwing away this zero mode works completely fine. Here's some pseudocode

```
elocs = jnp.matmul(evecs.conj().T,elocs)
elocs = jnp.where(evals/jnp.amax(evals) > rcond,evals/e,0)
elocs = jnp.matmul(evecs,elocs)
```

I find that `rcond` can be set to as low as 10^-12 on some problems and its usually fine at 10^-6
